### PR TITLE
#2032 Twilio Status Lookup and Update

### PR DESCRIPTION
**Pull request by mchlwellman**
# Description

This PR adds a few methods to the TwilioSMSClient, one to get the current message status from Twilio, and one to update an internal notification with the current Twilio status. The issue that this code is addressing is for when a callback doesn't get processed. A future ticket will add a celery beat task to update notifications from Twilio.

issue #2032

## How Has This Been Tested?

This was tested by adding some code to the generic internal endpoint, so that when a request was made to

``` http
GET /vanotify/internal/twilio?sid=xxx
```

the generic internal route handler called the `update_twilio_status` method with the `sid` from the request args.

Here are the test results,

- I updated the notification manually with psql:

```
notification_api=> update notifications set notification_status='sent' where reference='SMf9026475ad741347e0a1800cf9b98e66';
UPDATE 1
```

- I then triggered the Twilio status update by hitting the generic endpoint:
``` http
GET /vanotify/internal/twilio?sid=SMf9026475ad741347e0a1800cf9b98e66 HTTP/1.1
User-Agent: PostmanRuntime/7.42.0
Accept: */*
Host: dev-api.va.gov
 
HTTP/1.1 200 OK
Date: Tue, 22 Oct 2024 16:08:19 GMT
 
GET request received for endpoint /internal/twilio?sid=SMf9026475ad741347e0a1800cf9b98e66
```

- Checked in the db for the updated status:
Screenshot 2024-10-22 at 12 21 27 PM

- Reverted the generic internal changes and [redeployed successfully](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/11464492666).
## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an [appropriate title](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/team_agreement.md#naming-prs): #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Engineering/pull_request_labels.md#vanotify-labels) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the [style guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/style_guide.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works. [Testing guidelines](https://github.com/department-of-veterans-affairs/vanotify-team/blob/main/Process/testing_guide.md)
- [x] I have ensured the latest main is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] The ticket was moved into the DEV test column when I began testing this change

### CODE REVIEW

File: app/clients/sms/twilio.py [hunk 1]
```diff
@@ -232,22 +242,64 @@ def translate_delivery_status(
...
+            message_sid,
+            status,
+            status_reason,
+        )
+
+        self.logger.debug('Twilio delivery status translation: %s', status)
+
+        return status
+
+    def update_notification_status(self, message_sid: str) -> None:
```
Comment (cris-oddball) [inline app/clients/sms/twilio.py:264]: Should we log the notification_id here as well as the message SID?

File: app/clients/sms/twilio.py [hunk 2]
```diff
@@ -125,6 +129,12 @@ def name(self):
     def get_name(self):
         return self.name
 
+    def get_twilio_message(self, message_sid: str) -> MessageInstance | None:
```
Comment (cris-oddball) [inline app/clients/sms/twilio.py:132]: docstring?

File: app/clients/sms/twilio.py [hunk 3]
```diff
@@ -232,22 +242,64 @@ def translate_delivery_status(
...
         decoded_msg = base64.b64decode(twilio_delivery_status_message).decode()
-
         parsed_dict = parse_qs(decoded_msg)
-        message_sid = parsed_dict['MessageSid'][0]
-        twilio_delivery_status = parsed_dict['MessageStatus'][0]
-        error_code_data = parsed_dict.get('ErrorCode', None)
 
+        return decoded_msg, parsed_dict
+
+    def _evaluate_status(self, message_sid: str, twilio_delivery_status: str, error_codes: list) -> tuple[str, str]:
```
Comment (cris-oddball) [inline app/clients/sms/twilio.py:296]: docstring?

File: app/clients/sms/twilio.py (hunk 1 above)
Comment (mchlwellman) [reply app/clients/sms/twilio.py:273]: The whole thing is done by the message sid, the method doesn't have easy access to the notification id.

Comment (cris-oddball) [reply app/clients/sms/twilio.py:273]: I see, we update by referencing the `message_sid`

**Review by k-macmillan — APPROVED** [review]

**Review by cris-oddball — APPROVED** [review]

**Pull request merged**